sgkit/distance/metrics.py

```diff
@@ -172,3 +175,231 @@ def correlation_reduce(v: ArrayLike, out: ArrayLike) -> None:  # pragma: no cove
+
+    zero = types.uint32(0)
+
+    while i < m:
+        if x[i] >= zero and y[i] >= zero:
+            v0 += x[i]
+            v1 += y[i]
+            v2 += x[i] * x[i]
+            v3 += y[i] * y[i]
+            v4 += x[i] * y[i]
+            v5 += 1
+        i = types.uint32(i + types.uint32(1))
+
+    out[0] = v0
+    out[1] = v1
+    out[2] = v2
+    out[3] = v3
+    out[4] = v4
+    out[5] = v5
+
+
+@cuda.jit  # type: ignore
+def correlation_map_kernel(
+    x: ArrayLike, y: ArrayLike, out: ArrayLike
+) -> None:  # pragma: no cover.
+    i1 = types.uint32(cuda.grid(2)[types.uint32(0)])
+    i2 = types.uint32(cuda.grid(2)[types.uint32(1)])
+
+    out_shape_0 = types.uint32(out.shape[types.uint32(0)])
+    out_shape_1 = types.uint32(out.shape[types.uint32(1)])
+
+    if i1 >= out_shape_0 or i2 >= out_shape_1:
+        # Quit if (x, y) is outside of valid output array boundary
+        return
+
+    _correlation(x[i1], y[i2], out[i1][i2])
+
+
+def correlation_map_gpu(x: ArrayLike, y: ArrayLike) -> ArrayLike:  # pragma: no cover.
+    """Pearson correlation "map" function for partial vector pairs on GPU
+
+    Parameters
+    ----------
+    x
+        [array-like, shape: (m, n)]
+    y
+        [array-like, shape: (p, n)]
+
+    Returns
+    -------
+    An ndarray, which contains the output of the calculation of the application
+    of pearson correlation on the given pair of chunks, without the aggregation.
+    """
+
+    return call_metric_kernel(x, y, "correlation", correlation_map_kernel)
+
+
+def correlation_reduce_gpu(v: ArrayLike) -> None:  # pragma: no cover.
+    """GPU implementation of the corresponding "reduce" function for pearson
+    correlation.
+
+    Parameters
+    ----------
+    v
+        [array-like, shape: (1, n)]
+        The correlation array on which map step pearson corrections has been
+        applied.
+
+    Returns
+    -------
+    An ndarray, which contains the result of the calculation of the reduction step
+    of correlation metric.
+    """
+    return correlation_reduce_cpu(v)  # type: ignore
+
+
+@cuda.jit(device=True)  # type: ignore
+def _euclidean_distance_map(
+    a: ArrayLike, b: ArrayLike, out: ArrayLike
+) -> None:  # pragma: no cover.
+    """Helper function for the map step of euclidean distance which runs on
+    the device (GPU) itself.
+
+    Parameters
+    ----------
+    a
+        [array-like, shape: (1, n)]
+    b
+        [array-like, shape: (1, n)]
+    out
+        [array-like, shape: (1)]
+        The output array for returning the result.
+
+    Returns
+    -------
+    An ndarray, which contains the squared sum of the corresponding elements
+    of the given pair of vectors.
+    """
+    square_sum = types.float32(0)
+
+    zero = types.uint32(0)
+    a_shape_0 = types.uint32(a.shape[types.uint32(0)])
+    i = types.uint32(0)
+
+    while i < a_shape_0:
+        if a[i] >= zero and b[i] >= zero:
+            square_sum += (a[i] - b[i]) ** types.uint32(2)
```

Comment:
This was not intentional, it's just what I happened to use.
To be honest, I don't know the the root cause of why it's more performant to explicitly type everything apart from of course the obvious reason, that the compiler won't have to waste time to infer types.

Reply:
I don't mind if the compiler takes a tiny bit longer to compile the code - that seems worth having code without these inline typecasts on variable updates. Unless it significantly affects the actual run-time performance I'd prefer we didn't have them.

Reply:
It is not about compilation time, it's about runtime performance.

Reply:
Thanks @leofang - I'd like to understand what's going on here so I appreciate you taking the time to jump in here. I can see how we're helping out the compiler here to make good code in this case because it's an exponentiation. I don't understand the loop increment above though.

So, the question is if ``i`` is declared as a uint32, why we need to write ``i = types.uint32(i + types.uint32(1))`` rather than ``i += 1``?

I'm concerned because the patterns established in the PR will be copied in subsequent GPU kernels, and it does make the code quite a lot less easy to read. Declaring variables as particular types is totally fine, but  ``i = types.uint32(i + types.uint32(1))`` obscures the meaning of the code and makes mistakes much more likely.

Reply:
Numba infers the type of integer constants as being pointer-sized, i.e. 64-bit on most systems. This is point 2 in this section of the documentation: https://numba.readthedocs.io/en/latest/proposals/integer-typing.html#proposal-predictable-width-conserving-typing - the rationale for this typing is outlined in the document, but it can be summarised as "choose types that makes the typing behaviour seem as consistent and predictable as possible to the end user" - the previous ethos was "start small, and grow as required".

The impact of the current behaviour on the CUDA target is that judicious use of explicit types is needed, as in this line - if it were just `square_sum += (a[i] - b[i]) ** 2`, then `2` would be inferred as an `int64`, then the `**` would operate on `float32 ** int64`, which would then have both operands cast to use the `float64 ** float64` implementation of `**`, and then that would force `square_sum` to have the `float64` type everywhere, which is what slows things down.

[This document](https://github.com/numba/nvidia-cuda-tutorial/blob/main/session-2/session-2-with-notes.pdf) contains various examples and explanations relevant to the CUDA target.

Reply:
> So, the question is if `i` is declared as a uint32, why we need to write `i = types.uint32(i + types.uint32(1))` rather than `i += 1`?

It's not *declared* as a `uint32` - `i` refers to a `uint32` value at that point in the function, but if a value of a different type is stored in `i` somewhere else, then Numba has to unify the set of different types that `i` can store. This usually means using the biggest type that is used, or possibly something bigger than any of them - for example, if it stored an `int32` in one location and a `float32` on another, then the resulting unified type would be `float64`, because Numba tries to avoid losing precision (in this case, a `float32` doesn't have enough precision to represent all the digits of a large `int32` value).

Reply:
Great, thanks @gmarkall, that makes a lot of sense and the reference material is super helpful. So, in the loop counter example, if we wrote
```
i = types.uint32(0)
while stuff:
    i += 1
```
``i`` would be promoted to int64, because is 1 an int64 even though we've explicitly declared it to be a uint32?

So, then it should be sufficient to write
```
i = types.uint32(0)
while stuff:
    i += types.uint32(1)
```
and ``i = types.uint32(i + types.uint32(1))`` is excessive?

Reply:
> `i` would be promoted to int64, because is 1 an int64 even though we've explicitly declared `i` to be a uint32?

The code doesn't *explicitly declare* anything about `i` - it assigns it a `uint32` value on the first line. The type of the value it holds on the first line doesn't constrain the types of the values that might be assigned to it later. (for completeness / correctness, I would note though that using incompatible types can result in a unification error)



> So, then it should be sufficient to write
> 
> ```
> i = types.uint32(0)
> while stuff:
>     i += types.uint32(1)
> ```
> 
> and `i = types.uint32(i + types.uint32(1))` is excessive?

Unfortunately rule 3 of https://numba.readthedocs.io/en/latest/proposals/integer-typing.html#proposal-predictable-width-conserving-typing requires that you write `i = types.uint32(i + types.uint32(1))` - if you just write `i += types.uint32(1)`, the `+=` on two `uint32` values results in a `uint64`.

As an example of how to investigate this sort of thing, if you run:

```python
from numba import njit, types

@njit
def f(n):
    i = types.uint32(0)
    while i < types.uint32(n):
        i += types.uint32(1)
    return i

# Call the function so we get a compiled version to inspect
f(10)

# Inspect the result of Numba's type inference
f.inspect_types()
```

you will get a lot of output (the entire Numba IR and associated typing will be dumped) but the relevant parts are:

* The cast of the literal `0` down to `uint32`:
  ```
  #   i = call $4load_method.1($const6.2, func=$4load_method.1, args=[Var($const6.2, nocast.py:5)], kws=(), vararg=None, target=None)  :: (int64,) -> uint32
  ```
  * The phi node where the two different `i` values meet at the top of the loop (a phi node selects a variable from two different control flow paths - in this case there's an incoming `i` value from just before the while loop, and one from the loop body). This is a `uint64`, because one of the incoming values was a `uint64`:
  ```
  #   i.2 = phi(incoming_values=[Var(i, nocast.py:5), Var(i.1, nocast.py:7)], incoming_blocks=[0, 26])  :: uint64
  ```
* The cast of the literal `1` down to `uint32`:
  ```
  #   $34call_method.4 = call $30load_method.2($const32.3, func=$30load_method.2, args=[Var($const32.3, nocast.py:7)], kws=(), vararg=None, target=None)  :: (int64,) -> uint32
  ```
* The addition, which is typed as a `uint64`:
  ```
  #   $36inplace_add.5 = inplace_binop(fn=<built-in function iadd>, immutable_fn=<built-in function add>, lhs=i.2, rhs=$34call_method.4, static_lhs=Undefined, static_rhs=Undefined)  :: uint64
  ```
* The assignment of the result of the addition (a `uint64`) to `i` on the third line of the function (which must therefore also be a `uint64`):
  ```
      #   i.1 = $36inplace_add.5  :: uint64
  ```

Numba's typing works by propagating type information through the function until there are no more updates. You can see each step of the inference by running the code with `NUMBA_DEBUG_TYPEINFER=1` - each step is delimited by a banner stating `--- propagate ---`. The Numba IR and the current typing is then dumped at each step. Looking at the first step can often help pinpoint where widening is coming from. For the example above, the IR and first step look like:

```
$ NUMBA_DEBUG_TYPEINFER=1 python nocast.py 
---------------------------------SSA IR DUMP: f---------------------------------
label 0:
    n = arg(0, name=n)                       ['n']
    $2load_global.0 = global(types: <module 'numba.core.types' from '/home/gmarkall/numbadev/numba/numba/core/types/__init__.py'>) ['$2load_global.0']
    $4load_method.1 = getattr(value=$2load_global.0, attr=uint32) ['$2load_global.0', '$4load_method.1']
    $const6.2 = const(int, 0)                ['$const6.2']
    i = call $4load_method.1($const6.2, func=$4load_method.1, args=[Var($const6.2, nocast.py:5)], kws=(), vararg=None, target=None) ['$4load_method.1', '$const6.2', 'i']
    jump 12                                  []
label 12:
    i.2 = phi(incoming_values=[Var(i, nocast.py:5), Var(i.1, nocast.py:7)], incoming_blocks=[0, 26]) ['i', 'i.1', 'i.2']
    $14load_global.1 = global(types: <module 'numba.core.types' from '/home/gmarkall/numbadev/numba/numba/core/types/__init__.py'>) ['$14load_global.1']
    $16load_method.2 = getattr(value=$14load_global.1, attr=uint32) ['$14load_global.1', '$16load_method.2']
    $20call_method.4 = call $16load_method.2(n, func=$16load_method.2, args=[Var(n, nocast.py:5)], kws=(), vararg=None, target=None) ['$16load_method.2', '$20call_method.4', 'n']
    $22compare_op.5 = i.2 < $20call_method.4 ['$20call_method.4', '$22compare_op.5', 'i.2']
    bool24 = global(bool: <class 'bool'>)    ['bool24']
    $24pred = call bool24($22compare_op.5, func=bool24, args=(Var($22compare_op.5, nocast.py:6),), kws=(), vararg=None, target=None) ['$22compare_op.5', '$24pred', 'bool24']
    branch $24pred, 26, 42                   ['$24pred']
label 26:
    $28load_global.1 = global(types: <module 'numba.core.types' from '/home/gmarkall/numbadev/numba/numba/core/types/__init__.py'>) ['$28load_global.1']
    $30load_method.2 = getattr(value=$28load_global.1, attr=uint32) ['$28load_global.1', '$30load_method.2']
    $const32.3 = const(int, 1)               ['$const32.3']
    $34call_method.4 = call $30load_method.2($const32.3, func=$30load_method.2, args=[Var($const32.3, nocast.py:7)], kws=(), vararg=None, target=None) ['$30load_method.2', '$34call_method.4', '$const32.3']
    $36inplace_add.5 = inplace_binop(fn=<built-in function iadd>, immutable_fn=<built-in function add>, lhs=i.2, rhs=$34call_method.4, static_lhs=Undefined, static_rhs=Undefined) ['$34call_method.4', '$36inplace_add.5', 'i.2']
    i.1 = $36inplace_add.5                   ['$36inplace_add.5', 'i.1']
    jump 12                                  []
label 42:
    $44return_value.1 = cast(value=i.2)      ['$44return_value.1', 'i.2']
    return $44return_value.1                 ['$44return_value.1']

-----------------------------------propagate------------------------------------
---- type variables ----
[$14load_global.1 := Module(<module 'numba.core.types' from '/home/gmarkall/numbadev/numba/numba/core/types/__init__.py'>),
 $16load_method.2 := class(uint32),
 $20call_method.4 := uint32,
 $22compare_op.5 := bool,
 $24pred := bool,
 $28load_global.1 := Module(<module 'numba.core.types' from '/home/gmarkall/numbadev/numba/numba/core/types/__init__.py'>),
 $2load_global.0 := Module(<module 'numba.core.types' from '/home/gmarkall/numbadev/numba/numba/core/types/__init__.py'>),
 $30load_method.2 := class(uint32),
 $34call_method.4 := uint32,
 $36inplace_add.5 := uint64,
 $44return_value.1 := uint32,
 $4load_method.1 := class(uint32),
 $const32.3 := Literal[int](1),
 $const6.2 := Literal[int](0),
 arg.n := int64,
 bool24 := Function(<class 'bool'>),
 i := uint32,
 i.1 := uint64,
 i.2 := uint32,
 n := int64]
```

Numba IR is in Static Single Assignment (SSA) form, where variables are only assigned to once - the original Python variables are suffixed with an index to convert them to SSA form, so we have multiple instances of `i` - `i` (the original assignment), `i.1` (the assignment inside the loop), and `i.2` (the value of `i` at the start of the loop body coming either from outside the loop, or the end of a previous iteration). We can see that `i` started out holding a `uint32` (`i := uint32,`), as did `i.2` (` i.2 := uint32,`), but `i.1` held a `uint64` (`i.1 := uint64,`) because it was assigned the `uint64` result of the addition (` $36inplace_add.5 := uint64,`). In a subsequent step the `uint64` type of `i.1` propagated into `i.2` through the control flow from the end of the loop back to the top of it:

```
-----------------------------------propagate------------------------------------
... <snip> ...
 'i': uint32,
 'i.1': uint64,
 'i.2': uint64,
```

Finally, having investigated all this, we fix the problem with the cast of the result of the addition, so our code reads:

```python
from numba import njit, types

@njit
def f(n):
    i = types.uint32(0)
    while i < types.uint32(n):
        i = types.uint32(i + types.uint32(1))
    return i

print(f(10))
f.inspect_types()
```

then when we dump type inference debugging inference, the final types are (relevant info only, other bits omitted):

```
---------------------------------Variable types---------------------------------
... <snip > ...
 'i': uint32,
 'i.1': uint32,
 'i.2': uint32,
 'n': int64}
```

Reply:
Aha! Thank you so much @gmarkall, this is incredibly helpful. I am entirely satisfied :smile: 

Numba is, literally, awesome :star_struck:

Reply:
Apology @jeromekelleher, I was distracted and couldn't get back promptly, but looks like you are in good hands now 😅 Thanks @gmarkall for jumping in and providing such a detailed explanation! (I actually already linked to Graham's posts in my first reply in this PR.)